### ISSUE TITLE
Ubuntu - Documentation for Intel compiler build

### ISSUE THREAD

**Issue by ghost** [ethereum/webthree-umbrella#396]
Build System based on Ubuntu 14.04 and Intel Compiler 16.

Using these exact commits of the develop branches:
- CORESRC=/home/libweb3core
  fa1943c144121ae78b3f390a6a6fcdbeef59d953
- HELPERSRC=/home/webthree-helpers
  cf112efa34baa01a9cf60ec699537c840d34e9dd
- LIBSRC=/home/libethereum
  34b7273f69e98b789257dc03ad4fa6d89ba9b39c

Build with GNU compiler generates valid binaries.
Build with Intel Compiler generates boost::multiprecision Error

`/home/libweb3core/libdevcore/CommonData.h(120): error: no suitable conversion function from "dev::u256" to "uint8_t={unsigned char}" exists
        o_out[i - 1] = (typename Out::value_type)(uint8_t)v;
detected during instantiation of "void dev::toBigEndian(T, Out &) [with T=dev::u256, Out=std::string]" at line 137`

The build logs are attached.
Any suggestions?
[Ubuntu14.04_GNU_libethereum-build_bot-20160325182012.txt](https://github.com/ethereum/webthree-umbrella/files/190218/Ubuntu14.04_GNU_libethereum-build_bot-20160325182012.txt)
[Ubuntu14.04_INTEL_libethereum-build_bot-20160325183854.txt](https://github.com/ethereum/webthree-umbrella/files/190217/Ubuntu14.04_INTEL_libethereum-build_bot-20160325183854.txt)

**Comment by bobsummerwill** [comment]
Build with GCC?  :-)

But seriously, I assume that you have a specific reason for using the Intel compiler, but it looks like you are the first to try that particular combination and have hit some elements in our code (or its dependencies) which the Intel compiler doesn't like.

The "no suitable conversion ones" are happening for boost types, ie ...

using u256 =  boost::multiprecision::number<boost::multiprecision::cpp_int_backend<256, 256, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>>;

So you'll have to either work out for yourself what is happening there or seek help from the Boost team or Intel compiler team.    We're just using Boost, and it works on multiple platforms with multiple compilers, but maybe there is some shortfall for the Intel compiler?   I am afraid I must plead ignorance there.

Just spotted this in the log :-)   Which is quite true.   I'm afraid that the Intel compiler isn't something which we're claiming to support, but if you can get it to work, we'd be happy to take back pull-requests to add that support.

CMake Warning at webthree-helpers/cmake/EthCompilerSettings.cmake:93 (message):
  Your compiler is not tested, if you run into any issues, we'd welcome any
  patches.probl

There are a few more warnings in your log too, which it is hard to tell whether are serious or not.   I've just gone through the effort of fixing all our warnings and enabling warnings-as-errors, so we would want to get to the bottom of them all, and ensure we could build warning-free with the Intel compiler before pulling back changes.

Are you up for battling through them?    I'll do what I can to help, but that will mainly be cheerleading, I am afraid.   Good luck!

**Comment by ghost** [comment]
I will systematically resolve all issues to enable compilation with ICPC16.
Using the GNU binaries and the vanilla ethash_cl_miner_kernel.cl the k1om x100 5110P generates ~15Mhs.  
I expect that a native k1om application will outperform 15Mhs as the Xeon Phi supports 512bit SIMD instructions.

root@coast77:/home# /home/webthree-umbrella/build/libethereum/ethminer/ethminer --list-devices
[OPENCL]:
Listing OpenCL devices.
FORMAT: [deviceID] deviceName
[0] Intel(R) Many Integrated Core Acceleration Card
    CL_DEVICE_TYPE: ACCELERATOR
    CL_DEVICE_GLOBAL_MEM_SIZE: 6053646336
    CL_DEVICE_MAX_MEM_ALLOC_SIZE: 2017882112
    CL_DEVICE_MAX_WORK_GROUP_SIZE: 8192

**Comment by bobsummerwill** [comment]
Excellent.   All power to you!   Thanks.

**Comment by ghost** [comment]
Greetings, got most of the errors resolved using defines and convert_to.

```
#ifdef __INTEL_COMPILER
#else
#endif
```

```
.convert_to<uint8_t>()
.convert_to<unsigned>()
.convert_to<size_t>()
.convert_to<uint16_t>()
.convert_to<int>()
.convert_to<double>()
.convert_to<byte>()
.convert_to<string>()
.convert_to<uint64_t>()
```

Regarding \libweb3core\libdevcore\Base64.h:
The problem occurs with templates.

Do you have any advice for resolution of this particular case?

```


/home/libweb3core/libdevcore/Base64.h(48): error: no suitable conversion function from "boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<160U, 160U, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>, boost::multiprecision::et_off>" to "unsigned int" exists
        unsigned r = (unsigned)(a - a / 36 * 36); // boost's % is broken
                               ^
          detected during instantiation of "std::string dev::toBase36(const dev::FixedHash<N> &) [with N=20UL]" at line 114 of "/home/libethereum/libethcore/ICAP.cpp"


```

I tried convert_to, but got an error.

```
template <size_t N> inline std::string toBase36(FixedHash<N> const& _h)
{
    static char const* c_alphabet = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    typename FixedHash<N>::Arith a = _h;
    std::string ret;
    for (; a > 0; a /= 36)
    {
#ifdef __INTEL_COMPILER
        unsigned r = (unsigned)((a - a / 36 * 36).convert_to<unsigned>()); // boost's % is broken
#else
        unsigned r = (unsigned)(a - a / 36 * 36); // boost's % is broken
#endif
        ret = c_alphabet[r] + ret;
    }
    return ret;
}

```

```

/home/libweb3core/libdevcore/CommonData.h(121): error: type name is not allowed
        o_out[i - 1] = (typename Out::value_type)(uint8_t)(v.convert_to<uint8_t>());
                                                                        ^

/home/libweb3core/libdevcore/CommonData.h(121): error: expected an expression
        o_out[i - 1] = (typename Out::value_type)(uint8_t)(v.convert_to<uint8_t>());
                                                                                 ^

```

Regarding the template with toBigEndian, just enumerated like this, definitely requires rework.

```
/// Converts a templated integer value to the big-endian byte-stream represented on a templated collection.
/// The size of the collection object will be unchanged. If it is too small, it will not represent the
/// value properly, if too big then the additional elements will be zeroed out.
/// @a Out will typically be either std::string or bytes.
/// @a T will typically by unsigned, u160, u256 or bigint.
template <class T, class Out>
inline void toBigEndian(T _val, Out& o_out)
{
    static_assert(std::is_same<bigint, T>::value || !std::numeric_limits<T>::is_signed, "only unsigned types or bigint supported"); //bigint does not carry sign bit on shift
    for (auto i = o_out.size(); i != 0; _val >>= 8, i--)
    {
        T v = _val & (T)0xff;
        o_out[i - 1] = (typename Out::value_type)(uint8_t)v;
    }
}
#ifdef __INTEL_COMPILER
template <class Out>
inline void toBigEndian(u256 _val, Out& o_out)
{
    for (auto i = o_out.size(); i != 0; _val >>= 8, i--)
    {
        u256 v = _val & (u256)0xff;
        /* u256 

        o_out[i - 1] = (typename Out::value_type)(uint8_t)v;

        TODO:
        Resolve all cases cleanly
        */

        uint8_t vv;
        boost::multiprecision::number<boost::multiprecision::cpp_int_backend<256, 256, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>> bb;
        bb = (boost::multiprecision::number<boost::multiprecision::cpp_int_backend<256, 256, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>>)v;
        vv = (uint8_t)bb.convert_to<uint8_t>();
        o_out[i - 1] = (typename Out::value_type)vv;
    }
}
template <class Out>
inline void toBigEndian(u160 _val, Out& o_out)
{
    for (auto i = o_out.size(); i != 0; _val >>= 8, i--)
    {
        u160 v = _val & (u160)0xff;
        /* u160 

        o_out[i - 1] = (typename Out::value_type)(uint8_t)v;

        TODO:
        Resolve all cases cleanly
        */
        uint8_t vv;
        boost::multiprecision::number<boost::multiprecision::cpp_int_backend<160, 160, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>> bb;

        bb = (boost::multiprecision::number<boost::multiprecision::cpp_int_backend<160, 160, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>>)v;
        vv = (uint8_t)bb.convert_to<uint8_t>();
        o_out[i - 1] = (typename Out::value_type)vv;
    }
}
template <class Out>
inline void toBigEndian(bigint _val, Out& o_out)
{
    for (auto i = o_out.size(); i != 0; _val >>= 8, i--)
    {
        bigint v = _val & (bigint)0xff;
        /* bigint 

        o_out[i - 1] = (typename Out::value_type)(uint8_t)v;

        TODO:
        Resolve all cases cleanly
        */
        uint8_t vv;
        boost::multiprecision::number<boost::multiprecision::cpp_int_backend<>> bb;

        bb = (boost::multiprecision::number<boost::multiprecision::cpp_int_backend<>>)v;
        vv = (uint8_t)bb.convert_to<uint8_t>();
        o_out[i - 1] = (typename Out::value_type)vv;
    }
}
template <class Out>
inline void toBigEndian(boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<512U, 512U, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>, boost::multiprecision::et_off> _val, Out& o_out)
{
    for (auto i = o_out.size(); i != 0; _val >>= 8, i--)
    {
        boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<512U, 512U, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>, boost::multiprecision::et_off> v = _val & (boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<512U, 512U, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>, boost::multiprecision::et_off>)0xff;
        /* boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<512U, 512U, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>, boost::multiprecision::et_off> 

        o_out[i - 1] = (typename Out::value_type)(uint8_t)v;

        TODO:
        Resolve all cases cleanly
        */
        uint8_t vv;
        boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<512U, 512U, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>, boost::multiprecision::et_off> bb;

        bb = (boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<512U, 512U, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>, boost::multiprecision::et_off>)v;
        vv = (uint8_t)bb.convert_to<uint8_t>();
        o_out[i - 1] = (typename Out::value_type)vv;
    }
}
template <class Out>
inline void toBigEndian(boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<64U, 64U, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>, boost::multiprecision::et_off> _val, Out& o_out)
{
    for (auto i = o_out.size(); i != 0; _val >>= 8, i--)
    {
        boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<64U, 64U, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>, boost::multiprecision::et_off> v = _val & (boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<64U, 64U, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>, boost::multiprecision::et_off>)0xff;
        /* boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<64U, 64U, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>, boost::multiprecision::et_off> 

        o_out[i - 1] = (typename Out::value_type)(uint8_t)v;

        TODO:
        Resolve all cases cleanly
        */
        uint8_t vv;
        boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<64U, 64U, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>, boost::multiprecision::et_off> bb;

        bb = (boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<64U, 64U, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>, boost::multiprecision::et_off>)v;
        vv = (uint8_t)bb.convert_to<uint8_t>();
        o_out[i - 1] = (typename Out::value_type)vv;
    }
}
template <class Out>
inline void toBigEndian(boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<128U, 128U, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>, boost::multiprecision::et_off> _val, Out& o_out)
{
    for (auto i = o_out.size(); i != 0; _val >>= 8, i--)
    {
        boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<128U, 128U, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>, boost::multiprecision::et_off> v = _val & (boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<128U, 128U, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>, boost::multiprecision::et_off>)0xff;
        /* boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<128U, 128U, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>, boost::multiprecision::et_off> 

        o_out[i - 1] = (typename Out::value_type)(uint8_t)v;

        TODO:
        Resolve all cases cleanly
        */
        uint8_t vv;
        boost::multiprecision::number<boost::multiprecision::cpp_int_backend<>> bb;

        bb = (boost::multiprecision::number<boost::multiprecision::cpp_int_backend<>>)v;
        vv = (uint8_t)bb.convert_to<uint8_t>();
        o_out[i - 1] = (typename Out::value_type)vv;
    }
}
#endif
```

In order to minize build for ethminer binary, is it safe to disable Eth::evm and Eth::lll?
When trying to link ethminer:

```

[100%] Building CXX object libethereum/ethminer/CMakeFiles/ethminer.dir/main.cpp.o
Linking CXX executable ethminer
../libethereum/libethereum.so: undefined reference to `dev::eth::ExtVMFace::ExtVMFace(dev::eth::EnvInfo const&, dev::FixedHash<20u>, dev::FixedHash<20u>, dev::FixedHash<20u>, boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void>, (boost::multiprecision::expression_template_option)0>, boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)0, (boost::multiprecision::cpp_int_check_type)0, void>, (boost::multiprecision::expression_template_option)0>, dev::vector_ref<unsigned char const>, std::vector<unsigned char, std::allocator<unsigned char> >, dev::FixedHash<32u> const&, unsigned int)'
../libethereum/libethereum.so: undefined reference to `dev::eth::compileLLL(std::string const&, bool, std::vector<std::string, std::allocator<std::string> >*)'
../libethereum/libethereum.so: undefined reference to `dev::eth::VMFactory::create()'
../libethereum/libethereum.so: undefined reference to `dev::eth::VMFactory::create(dev::eth::VMKind)'
../libethereum/libethereum.so: undefined reference to `dev::eth::instructionInfo(dev::eth::Instruction)'
make[2]: *** [libethereum/ethminer/ethminer] Error 1
```

[Base64Error_libethereum-build_bot-20160327024444.txt](https://github.com/ethereum/webthree-umbrella/files/192943/Base64Error_libethereum-build_bot-20160327024444.txt)
[Base64_fail_libethereum-build_bot-20160329004639.txt](https://github.com/ethereum/webthree-umbrella/files/192942/Base64_fail_libethereum-build_bot-20160329004639.txt)
[Linker_libethereum-build_bot-20160329012817.txt](https://github.com/ethereum/webthree-umbrella/files/192944/Linker_libethereum-build_bot-20160329012817.txt)

**Comment by bobsummerwill** [comment]
Hey @mancoast!
Sorry for the lag ..

I would ignore ethminer binary if I were you.   For the time being at least.

Why don't you do a fork-and-clone and create some PRs for the changes in process which we can discuss?

**Comment by bobsummerwill** [comment]
For the type conversion issues, perhaps @chriseth will know a little better than me?

**Comment by chriseth** [comment]
Oh, this looks quite horrible. I have seen the following in the boost documentation, perhaps that helps:
`int i = z.template convert_to<int>(); // sets i to 2`

**Comment by bobsummerwill** [comment]
Yikes to that syntax!    Did it get you further on?

**Comment by ghost** [comment]
Yea, all the requirements have been fulfilled with the single exception of Base64.h.
https://github.com/pwntoast512/libethereum
https://github.com/pwntoast512/libweb3core

After completion of toBase36 I will fork using this username for a cleaner, single commit, pull request.
@bobsummerwill @chriseth 
Please advise fulfillment of the following requirement:

```
template <size_t N> inline std::string toBase36(FixedHash<N> const& _h)
{
    static char const* c_alphabet = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    typename FixedHash<N>::Arith a = _h;
    std::string ret;
    for (; a > 0; a /= 36)
    {
#ifdef __INTEL_COMPILER
        //TODO
        //unsigned r = (unsigned)((a - a / 36 * 36).convert_to<unsigned>()); // boost's % is broken
        unsigned r = 0;
#else
        unsigned r = (unsigned)(a - a / 36 * 36); // boost's % is broken
#endif
        ret = c_alphabet[r] + ret;
    }
    return ret;
}
```

Thanks!

**Comment by bobsummerwill** [comment]
"Please advise fulfillment of the following requirement:"

Sorry - I don't know what exactly you mean.    Please could you rephrase?   There would need to be some working expression inside the Intel #ifdef, wouldn't there?   Or the code would be broken.

**Comment by ghost** [comment]
Restored dev::toBase36 to vanilla && resolved, via ICAP modification.

```
#ifdef __INTEL_COMPILER
        static char const* c_alphabet = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        typename FixedHash<Address::size>::Arith a = m_direct;
        std::string dd;
        for (; a > 0; a /= 36)
        {
            unsigned r = (unsigned)((a - a / 36 * 36)).convert_to<unsigned>(); // boost's % is broken
            dd = c_alphabet[r] + dd;
        }
        std::string d = dd;
#else
        std::string d = toBase36<Address::size>(m_direct);
#endif
```

**Comment by bobsummerwill** [comment]
Have you still got outstanding breaks, or have you got to the link step yet?

**Comment by ghost** [comment]
ICC issue is resolved, on to the next one: Xeon PHI x100 Co-Processor
i am currently recompiling all required libraries for k1om, might be that simple...

**Comment by ghost** [comment]
![this_ones_trouble](https://cloud.githubusercontent.com/assets/7040493/14223186/03bfdf48-f849-11e5-8022-9f806f63e216.png)
[libethereum-build_bot-20160401193304.txt](https://github.com/ethereum/webthree-umbrella/files/200593/libethereum-build_bot-20160401193304.txt)
[ethmin-trials-20160401200517.txt](https://github.com/ethereum/webthree-umbrella/files/200594/ethmin-trials-20160401200517.txt)

Accidentally clicked the close button.
Binaries are confirmed functional with x64.

Looks like dual E5-2603 are hashing at ~.25 to .30 Mhs.
Code is functional, but requires cleanup.

**Comment by bobsummerwill** [comment]
Great - getting there, eh!
Thanks for all your work on this.   I just commented on one of the PRs.
I'd like to try to eliminate most of the #ifdefs if at all possible, to make it easier to maintain.   And if that means changes to the existing code, that is fine.

**Comment by ghost** [comment]
Greetings!

Questions:
1.) should i close the pull requests?
2.) How to handle scenarios that still require different logic for the ICPC [still need pound defines]?
3.) is pull request with multiple commits acceptable?

Please advise this decision point: How to proceed: a.) or b.)?
a.) add new commits to pull request
b.) close pull request, generate new diff, open new pull request.

Thanks,
coast

**Comment by bobsummerwill** [comment]
Just keep making further commits on those branches until we are both happy with the content.
The usual flow is to keep going with that, and then "squash" them just before committing.

Like http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

**Comment by ghost** [comment]
Alrighty, most of the defines have been removed.
The use of member function convert_to remains an issue with template parameters.
How should we proceed?

**Comment by bobsummerwill** [comment]
Just adding links to the two PRs:

https://github.com/ethereum/libethereum/pull/223
https://github.com/ethereum/libweb3core/pull/67

**Comment by bobsummerwill** [comment]
Just a few more questions.
1. For the toBigEndian() functions, are those OVERRIDES to existing functions?   Where are they called from?   Without those, what build errors do you get?   What fails?    I'm sure that what you have is correct, but I would just like to better understand where they are used.
2. For the toBase36 one in libethereum, should that actually be an override to that function within an ifdef in the base64.h in libdevcore?   rather than inlined in the call site?    Or would that not work?

Beyond that I would just want to check with @chriseth on the test run results we are seeing for these 2 PRs, which I think are issues in our automation, and nothing to do with your changes, and then I think we would be good to go.

No doubt further breaks will pop up in the future which we would be blind to, but obviously feel free to submit PRs as we go.

Did you get the performance improvements you were hoping for?

**Comment by bobsummerwill** [comment]
Oh - and the squashes will still need doing, but we would defer that until we're 100% happy with the content.    Thanks again for your persistence!

**Comment by ghost** [comment]
toBigEndian: T=u256 || T=u160

```
/home/cpp-ethereum/libdevcore/../libdevcore/CommonData.h(111): error: no suitable conversion function from "dev::u256" to "uint8_t={unsigned char}" exists
        o_out[i - 1] = (typename Out::value_type)(uint8_t)v;
                                                          ^
          detected during instantiation of "void dev::toBigEndian(T, Out &) [with T=dev::u256, Out=std::string]" at line 128

/home/cpp-ethereum/libdevcore/../libdevcore/CommonData.h(111): error: no suitable conversion function from "dev::u160" to "uint8_t={unsigned char}" exists
        o_out[i - 1] = (typename Out::value_type)(uint8_t)v;
                                                          ^
          detected during instantiation of "void dev::toBigEndian(T, Out &) [with T=dev::u160, Out=std::string]" at line 129

/home/cpp-ethereum/libdevcore/../libdevcore/CommonData.h(111): error: no suitable conversion function from "dev::u256" to "uint8_t={unsigned char}" exists
        o_out[i - 1] = (typename Out::value_type)(uint8_t)v;
                                                          ^
          detected during instantiation of "void dev::toBigEndian(T, Out &) [with T=dev::u256, Out=dev::bytes]" at line 130

/home/cpp-ethereum/libdevcore/../libdevcore/CommonData.h(111): error: no suitable conversion function from "dev::u160" to "uint8_t={unsigned char}" exists
        o_out[i - 1] = (typename Out::value_type)(uint8_t)v;
                                                          ^
          detected during instantiation of "void dev::toBigEndian(T, Out &) [with T=dev::u160, Out=dev::bytes]" at line 131

compilation aborted for /home/cpp-ethereum/libdevcore/FileSystem.cpp (code 2)
```

toBigEndian: T=bigint

```
/home/libweb3core/libdevcore/CommonData.h(120): error: no suitable conversion function from "dev::bigint" to "uint8_t={unsigned char}" exists
        o_out[i - 1] = (typename Out::value_type)(uint8_t)v;
                                                          ^
          detected during:
            instantiation of "void dev::toBigEndian(T, Out &) [with T=dev::bigint, Out=dev::bytes]" at line 195
            instantiation of "dev::bytes dev::toCompactBigEndian(T, unsigned int) [with T=dev::bigint]" at line 125 of "/home/libweb3core/rlp/main.cpp"

compilation aborted for /home/libweb3core/rlp/main.cpp (code 2)

```

toBigEndian: T=cpp_int_backend<512U>

```

/home/libweb3core/libdevcore/CommonData.h(120): error: no suitable conversion function from "boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<512U, 512U, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>, boost::multiprecision::et_off>" to "uint8_t={unsigned char}" exists
        o_out[i - 1] = (typename Out::value_type)(uint8_t)v;
                                                          ^
          detected during:
            instantiation of "void dev::toBigEndian(T, Out &) [with T=boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<512U, 512U, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>, boost::multiprecision::et_off>, Out=std::array<byte={uint8_t={unsigned char}}, 64UL>]" at line 72 of "/home/libweb3core/libdevcore/FixedHash.h"
            instantiation of "dev::FixedHash<N>::FixedHash(const dev::FixedHash<N>::Arith &) [with N=64U]" at line 91 of "/home/libweb3core/libdevcore/CommonJS.h"
            instantiation of "dev::FixedHash<N> dev::jsToFixed<N>(const std::string &) [with N=64U]" at line 37 of "/home/libethereum/libethcore/CommonJS.h"

compilation aborted for /home/libethereum/libethcore/CommonJS.cpp (code 2)

```

toBigEndian: T=cpp_int_backend<64U>

```

/home/libweb3core/libdevcore/CommonData.h(120): error: no suitable conversion function from "boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<64U, 64U, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>, boost::multiprecision::et_off>" to "uint8_t={unsigned char}" exists
        o_out[i - 1] = (typename Out::value_type)(uint8_t)v;
                                                          ^
          detected during:
            instantiation of "void dev::toBigEndian(T, Out &) [with T=boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<64U, 64U, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>, boost::multiprecision::et_off>, Out=std::array<byte={uint8_t={unsigned char}}, 8UL>]" at line 72 of "/home/libweb3core/libdevcore/FixedHash.h"
            instantiation of "dev::FixedHash<N>::FixedHash(const dev::FixedHash<N>::Arith &) [with N=8U]" at line 59 of "/home/libweb3core/libdevcore/Base64.h"
            instantiation of "dev::FixedHash<N> dev::fromBase36<N>(const std::string &) [with N=8UL]" at line 151 of "/home/libethereum/libethcore/ICAP.cpp"

compilation aborted for /home/libethereum/libethcore/ICAP.cpp (code 2)
```

toBigEndian: T=cpp_int_backend<128U>

```
/home/libweb3core/libdevcore/CommonData.h(120): error: no suitable conversion function from "boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<128U, 128U, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>, boost::multiprecision::et_off>" to "uint8_t={unsigned char}" exists
        o_out[i - 1] = (typename Out::value_type)(uint8_t)v;
                                                          ^
          detected during:
            instantiation of "void dev::toBigEndian(T, Out &) [with T=boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<128U, 128U, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>, boost::multiprecision::et_off>, Out=std::array<byte={uint8_t={unsigned char}}, 16UL>]" at line 72 of "/home/libweb3core/libdevcore/FixedHash.h"
            instantiation of "dev::FixedHash<N>::FixedHash(const dev::FixedHash<N>::Arith &) [with N=16U]" at line 735 of "/home/libethereum/ethkey/KeyAux.h"

compilation aborted for /home/libethereum/ethkey/main.cpp (code 2)
```

toBase36: initial error

```

[ 44%] Building CXX object libethereum/libethcore/CMakeFiles/ethcore.dir/ICAP.cpp.o
/home/libweb3core/libdevcore/Base64.h(48): error: no suitable conversion function from "boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<160U, 160U, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>, boost::multiprecision::et_off>" to "unsigned int" exists
        unsigned r = (unsigned)(a - a / 36 * 36); // boost's % is broken
                               ^
          detected during instantiation of "std::string dev::toBase36(const dev::FixedHash<N> &) [with N=20UL]" at line 118 of "/home/libethereum/libethcore/ICAP.cpp"

compilation aborted for /home/libethereum/libethcore/ICAP.cpp (code 2)
```

toBase36: member function fail

```
/home/libweb3core/libdevcore/Base64.h(49): error: type name is not allowed
        unsigned r = (unsigned)((a - a / 36 * 36).convert_to<unsigned>()); // boost's % is broken
                                                             ^

/home/libweb3core/libdevcore/Base64.h(49): error: expected an expression
        unsigned r = (unsigned)((a - a / 36 * 36).convert_to<unsigned>()); // boost's % is broken
                                                                       ^

compilation aborted for /home/libethereum/libethcore/ICAP.cpp (code 2)
```

I would prefer to use an template "override".
Is this different from C++11 override?
http://en.cppreference.com/w/cpp/language/override

What is the exact signature required for the additional toBase36 function?

I have tried using 20UL, but the function remains uncalled and icpc throws an error: no suitable conversion.

```
#ifdef __INTEL_COMPILER
inline std::string toBase36(FixedHash<20UL> const& _h)
{
    static char const* c_alphabet = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    typename FixedHash<20UL>::Arith a = _h;
    std::string ret;
    for (; a > 0; a /= 36)
    {
        unsigned r = (unsigned)((a - a / 36 * 36)).convert_to<unsigned>(); // boost's % is broken
        ret = c_alphabet[r] + ret;
    }
    return ret;
}
#endif
```

```

/home/libweb3core/libdevcore/Base64.h(48): error: no suitable conversion function from "boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<160U, 160U, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>, boost::multiprecision::et_off>" to "unsigned int" exists
        unsigned r = (unsigned)(a - a / 36 * 36); // boost's % is broken
                               ^
          detected during instantiation of "std::string dev::toBase36(const dev::FixedHash<N> &) [with N=20UL]" at line 118 of "/home/libethereum/libethcore/ICAP.cpp"

compilation aborted for /home/libethereum/libethcore/ICAP.cpp (code 2)
```

**Comment by ghost** [comment]
The baseline native performance for mic, requires all the libraries recompiled for mic.
Currently completed is gmp6_1 and boost1_55.
Many more libraries remain to be cross compiled before the ethereum binaries will link.

After establishing a baseline, it will be interesting to tune the ethminer binary.

**Comment by bobsummerwill** [comment]
If you are only on boost 55, it might be worth your while updating that.

1.60 is the latest version.

There might be a case for sending them some patches too, eh?

**Comment by bobsummerwill** [comment]
I'm still not clear on where toBigEndian() comes from.    Is that boost API?

**Comment by ghost** [comment]
I tested with boost 1-60 release, but there was no resolution. My guess is that the Intel compiler has more strict lexi. I would not consider that unexpected from Intel platform.  As for additional boost patch, yes the "using bigint =.." causes another error. That is separate scope, but will be analysed later. 

ToBigEndian is a template core function of libweb3core, not related to boost. The error occurs when this libweb3core template function is instantiated with boost's wide types.

**Comment by bobsummerwill** [comment]
Yes, I meant an overload template function, not the "override" keyword added in C++11.

Tell you what, why don't we break up the problem set a bit here?

Could you create new PRs for the two libraries containing ONLY the convert_to<> and pragma changes?   We can commit those right now.

Then maybe one PR for the toBase36, one for the toBigEndian() functions and one for the pushInt(), and then we can work through these issues peacemeal, using our collective brainpower to try to come up with the most minimal ways of adding this support.

Getting the mass of those easy changes in reduces the diffs to consider.

Sound like a plan?   Thanks!

**Comment by ghost** [comment]
sounds like a solid plan of action.
I will minimize changes in 223 and 67 to include ONLY the convert_to<> and pragma changes.
ethereum/libethereum#223
ethereum/libweb3core#67
Then create a new pull request for the template functions, etc..

**Comment by bobsummerwill** [comment]
Cool.

BTW - good news on squashing!   See https://twitter.com/BobSummerwill/status/716773401675714560.

**Comment by bobsummerwill** [comment]
OK ... so I have committed those two.  Thanks!

Something I noticed after committing which it would good to address is that the warning pragmas don't have push/pops around them.   Not sure if the Intel compiler supports that?

But best practice on such pragmas (if compiler supports) is to pattern them like so ...

```
#ifdef SPECIFIC_COMPILER
#pragma warning(push)
#pragma warning(disable:1234)
#endif

// Specifically targeted "bad code", whether that is our code, or external code
// Obviously seeking to minimize this as much as possible (which can even mean
// nesting these pragmas inside a certain function, just around a specific bad block.

#ifdef SPECIFIC_COMPILER
#pragma warning(pop)
#endif
```

With that pattern, the "loose area" just continues to end of compilation unit.    I know there are some other modes in there, like restore to "default", but it's only really push/pop which has correct semantics.

Not sure if those pragmas are part of the C++ standard and supported by all compliant compilers?

**Comment by ghost** [comment]
Regarding the pragmas, this requires addition analysis.

I would like to keep develop branch of https://github.com/mancoast/libethereum synced with the develop branch from https://github.com/ethereum/libethereum.

How to proceed? a.) or b.)
a.) git reset https://github.com/mancoast/libethereum to a common commit, then pull from https://github.com/ethereum/libethereum.
b.) merge branches like this: generates an additional merge commit
![image](https://cloud.githubusercontent.com/assets/7040493/14262735/ec611180-fa84-11e5-9059-0d69c04abc09.png)

Thanks,
coast

**Comment by bobsummerwill** [comment]
(b) is just fine.

We've got the squash-as-part-of-commit functionality in Github now, so we don't need to bother about manually squashing or minimizing commits.    Those "merge remote tracking" commits don't seem to appear on the other side of pull-request merges anyway, but having multiple commits in a PR, some of which will make it through and some of which won't certainly can be a little confusion.

But yeah - all moot anyway now.

**Comment by chriseth** [comment]
The problem with merge commits is that it is impossible to review conflict resolution in github, so please still rebase. @bobsummerwill

**Comment by ghost** [comment]
Ok I played around and got everything squared away with the sync.
Thanks for helping me get accustomed to this version control system.

**Comment by bobsummerwill** [comment]
no problem.   Give me a shout when you have more PRs ready for review, and we can start taking our next steps.

**Comment by ghost** [comment]
@bobsummerwill 
https://github.com/ethereum/libweb3core/pull/68

Figured out the correct signature for an overloaded toBase36 function.
Still need to confirm the build.

**Comment by bobsummerwill** [comment]
Right - thanks.

Myself and the rest of the team are still trying to understand the new PR as well, to ensure that we commit any changes intentionally, and with a plan for not needing to special case the Intel-compiler in the future when the root causes are fixed.     Or, if our code is wrong and needs changing, to do that instead.

**Comment by ghost** [comment]
The implicit conversion used in the template function does not compile using Intel platform.

```
u256 test = 0;
unsigned testing = (unsigned)test; // will not compile using Intel Platform
```

With the argument type specified as multi precision boost, we can call the convert_to member function.

```
u256 test = 0;
unsigned testing = test.convert_to<unsigned>(); // will compile using Intel Platform
```

In summary, the intel platform supports this "convert_to" member function, but does not support implicit conversion of boost multiprecision types

```
template <class T, class Out>
inline void toBigEndian(T _val, Out& o_out)
{
    for (auto i = o_out.size(); i != 0; _val >>= 8, i--)
    {
        T v = _val & (T)0xff; // could be wide boost, could be int, both are valid
        o_out[i - 1] = (typename Out::value_type)(uint8_t)v; // this will compile if T=int
                     // it will break if T=u256
    }
}
```

This "convert_to" member function is not available with the template parameter type, because the template variable could be any type.

```
template <class T, class Out>
inline void toBigEndian(T _val, Out& o_out)
{
    for (auto i = o_out.size(); i != 0; _val >>= 8, i--)
    {
        T v = _val & (T)0xff; // could be wide boost, could be int, both are valid
        o_out[i - 1] = (typename Out::value_type)v.convert_to<uint8_t>(); 
              // this will not compile ever
    }
}

```

This issue stems from lossy implicit conversion using boost multi-precision types.  GCC allows implicitly lossy conversion with boost multi-precision types, Intel does not.  The solution is to prevent the template function instantiation for boost multi-precision types via function overloading.

```
#ifdef __INTEL_COMPILER
template <class Out>
inline void toBigEndian(u256 _val, Out& o_out)
{
    for (auto i = o_out.size(); i != 0; _val >>= 8, i--)
    {
        o_out[i - 1] = (typename Out::value_type)(_val & (u256)0xff).convert_to<uint8_t>();
    }
}
#endif
```

**Comment by chriseth** [comment]
@mancoast as the "EthCompilerSettings" file does not know the intel compiler, it does not set any flags. Did you modify it accordingly? Which flags do you use?

**Comment by bobsummerwill** [comment]
@mancoast How do you install the Intel compiler for Ubuntu?
I would like to replicate your setup.

**Comment by ghost** [comment]
Greetings @bobsummerwill

Intel compiler installer supports Ubuntu using the supplied install_GUI.sh out of the box. 
I have recently switched back to CentOS7.  everything works like a charm, regardless of OS when compiling just libweb3core and libethereum. However, I aim only to compile the ethminer binary for k1om.

**Comment by bobsummerwill** [comment]
Could you share a link to the correct Intel Compiler install, please, @mancoast?
Their website is confusing.    Not sure where to look for the right now.   Thanks!

And just to confirm - you aren't hindered by the current state of things, right?    That is enough for ethminer to build?    Or are you still hindered, and the unmerged ifdefs were just sufficient for a "quick win" for the rest of the code?

**Comment by ghost** [comment]
Hello,

The unmerged ifdefs are required to build ethminer.  

To use Intel compiler, download the trial of Intel Parallel Studio XE:
https://software.intel.com/en-us/intel-parallel-studio-xe

**Comment by ghost** [comment]
Greetings @bobsummerwill @chriseth,

We are very close to Xeon Phi native execution for ethminer.
https://github.com/ethereum/libweb3core/pull/71
Attached is the build-log - [libethereum-build_bot-20160429075256.txt](https://github.com/ethereum/webthree-umbrella/files/242400/libethereum-build_bot-20160429075256.txt)

All libraries compile with -mmic.
However, there exist some link errors for ethminer.

In particular, the cryptoPP references require additional rework because these functions depend on SSE2.  SSE2 is not supported on k1om Xeon Phi.

```
../../libweb3core/libp2p/libp2p.so: undefined reference to `CryptoPP::AlignedDeallocate(void*)'
../../libweb3core/libp2p/libp2p.so: undefined reference to `CryptoPP::AlignedAllocate(unsigned long)'
```

I can work through most of the other errors:

```
/opt/intel/compilers_and_libraries_2016.2.181/linux/bin/intel64/icpc   -std=c++11 -mmic    -mmic CMakeFiles/ethminer.dir/main.cpp.o  -o ethminer -rdynamic ../../../../jsoncpp-1.6.2/build-mic/src/lib_json/libjsoncpp.a ../../../../json-rpc-cpp-0.5.0/build-mic/lib/libjsonrpccpp-common.so ../../../../json-rpc-cpp-0.5.0/build-mic/lib/libjsonrpccpp-client.so ../../../../mpss-3.7/k1om/usr/lib64/libcurl.so ../../../../jsoncpp-1.6.2/build-mic/src/lib_json/libjsoncpp.a ../../../../leveldb-1.3/build-mic/libleveldb.a ../../../../boost_1_55_0/build-mic/lib/libboost_thread.so ../../../../boost_1_55_0/build-mic/lib/libboost_random.so ../../../../boost_1_55_0/build-mic/lib/libboost_filesystem.so ../../../../boost_1_55_0/build-mic/lib/libboost_system.so -lpthread ../../../../cryptopp-5.6.2/build-mic/lib/libcryptopp.a ../../webthree-helpers/utils/secp256k1/libsecp256k1.a ../../../../boost_1_55_0/build-mic/lib/libboost_regex.so ../../../../cryptopp-5.6.2/build-mic/lib/libcryptopp.a ../libethashseal/libethashseal.so ../libethereum/libethereum.so ../libevm/libevm.so ../libethcore/libethcore.so ../libevmcore/libevmcore.so ../../libweb3core/libp2p/libp2p.so ../../libweb3core/libdevcrypto/libdevcrypto.so ../../libweb3core/libdevcore/libdevcore.so ../../../../jsoncpp-1.6.2/build-mic/src/lib_json/libjsoncpp.a ../../../../leveldb-1.3/build-mic/libleveldb.a ../../../../boost_1_55_0/build-mic/lib/libboost_thread.so ../../../../boost_1_55_0/build-mic/lib/libboost_random.so ../../../../boost_1_55_0/build-mic/lib/libboost_filesystem.so ../../../../boost_1_55_0/build-mic/lib/libboost_system.so -lpthread ../../webthree-helpers/utils/libscrypt/libscrypt.so ../../webthree-helpers/utils/secp256k1/libsecp256k1.a ../../../build-gmp/gmp/mic/lib/libgmp.a ../../../../boost_1_55_0/build-mic/lib/libboost_regex.so ../libethash/libethash.so ../../../../cryptopp-5.6.2/build-mic/lib/libcryptopp.a -Wl,-rpath,/home/json-rpc-cpp-0.5.0/build-mic/lib:/home/mpss-3.7/k1om/usr/lib64:/home/boost_1_55_0/build-mic/lib:/home/libethereum-buildbot/build-mic/libethereum/libethashseal:/home/libethereum-buildbot/build-mic/libethereum/libethereum:/home/libethereum-buildbot/build-mic/libethereum/libevm:/home/libethereum-buildbot/build-mic/libethereum/libethcore:/home/libethereum-buildbot/build-mic/libethereum/libevmcore:/home/libethereum-buildbot/build-mic/libweb3core/libp2p:/home/libethereum-buildbot/build-mic/libweb3core/libdevcrypto:/home/libethereum-buildbot/build-mic/libweb3core/libdevcore:/home/libethereum-buildbot/build-mic/webthree-helpers/utils/libscrypt:/home/libethereum-buildbot/build-mic/libethereum/libethash: 
x86_64-k1om-linux-ld: warning: libimf.so, needed by ../../../../json-rpc-cpp-0.5.0/build-mic/lib/libjsonrpccpp-common.so, not found (try using -rpath or -rpath-link)
x86_64-k1om-linux-ld: warning: libsvml.so, needed by ../../../../json-rpc-cpp-0.5.0/build-mic/lib/libjsonrpccpp-common.so, not found (try using -rpath or -rpath-link)
x86_64-k1om-linux-ld: warning: libirng.so, needed by ../../../../json-rpc-cpp-0.5.0/build-mic/lib/libjsonrpccpp-common.so, not found (try using -rpath or -rpath-link)
x86_64-k1om-linux-ld: warning: libintlc.so.5, needed by ../../../../json-rpc-cpp-0.5.0/build-mic/lib/libjsonrpccpp-common.so, not found (try using -rpath or -rpath-link)
x86_64-k1om-linux-ld: warning: libcap.so.2, needed by ../../../../mpss-3.7/k1om/usr/lib64/libcurl.so, not found (try using -rpath or -rpath-link)
../../libweb3core/libp2p/libp2p.so: undefined reference to `CryptoPP::AlignedDeallocate(void*)'
/home/mpss-3.7/k1om/usr/lib64/libgcrypt.so.11: undefined reference to `cap_from_text'
/home/mpss-3.7/k1om/usr/lib64/libgcrypt.so.11: undefined reference to `cap_set_proc'
../../libweb3core/libp2p/libp2p.so: undefined reference to `CryptoPP::AlignedAllocate(unsigned long)'
make[2]: *** [libethereum/ethminer/ethminer] Error 1
make[2]: Leaving directory `/home/libethereum-buildbot/build-mic'
make[1]: *** [libethereum/ethminer/CMakeFiles/ethminer.dir/all] Error 2
make[1]: Leaving directory `/home/libethereum-buildbot/build-mic'
make: *** [all] Error 2
```

Thanks,
coast

**Comment by ghost** [comment]
Manually modify rpath:
`-Wl,-rpath,[...]:/opt/intel/compilers_and_libraries_2016.2.181/linux/compiler/lib/intel64_lin_mic/:/home/mpss-3.7/k1om/lib64/`

Fixes most of the link errors, except the CryptoPP references.
Any suggestions?

```
[root@coast77 ethminer]# cd /home/libethereum-buildbot/build-mic/libethereum/ethminer && /opt/intel/compilers_and_libraries_2016.2.181/linux/bin/intel64/icpc   -std=c++11 -mmic    -mmic CMakeFiles/ethminer.dir/main.cpp.o  -o ethminer -rdynamic ../../../../jsoncpp-1.6.2/build-mic/src/lib_json/libjsoncpp.a ../../../../json-rpc-cpp-0.5.0/build-mic/lib/libjsonrpccpp-common.so ../../../../json-rpc-cpp-0.5.0/build-mic/lib/libjsonrpccpp-client.so ../../../../mpss-3.7/k1om/usr/lib64/libcurl.so ../../../../jsoncpp-1.6.2/build-mic/src/lib_json/libjsoncpp.a ../../../../leveldb-1.3/build-mic/libleveldb.a ../../../../boost_1_55_0/build-mic/lib/libboost_thread.so ../../../../boost_1_55_0/build-mic/lib/libboost_random.so ../../../../boost_1_55_0/build-mic/lib/libboost_filesystem.so ../../../../boost_1_55_0/build-mic/lib/libboost_system.so -lpthread ../../../../cryptopp-5.6.2/build-mic/lib/libcryptopp.a ../../webthree-helpers/utils/secp256k1/libsecp256k1.a ../../../../boost_1_55_0/build-mic/lib/libboost_regex.so ../../../../cryptopp-5.6.2/build-mic/lib/libcryptopp.a ../libethashseal/libethashseal.so ../libethereum/libethereum.so ../libevm/libevm.so ../libethcore/libethcore.so ../libevmcore/libevmcore.so ../../libweb3core/libp2p/libp2p.so ../../libweb3core/libdevcrypto/libdevcrypto.so ../../libweb3core/libdevcore/libdevcore.so ../../../../jsoncpp-1.6.2/build-mic/src/lib_json/libjsoncpp.a ../../../../leveldb-1.3/build-mic/libleveldb.a ../../../../boost_1_55_0/build-mic/lib/libboost_thread.so ../../../../boost_1_55_0/build-mic/lib/libboost_random.so ../../../../boost_1_55_0/build-mic/lib/libboost_filesystem.so ../../../../boost_1_55_0/build-mic/lib/libboost_system.so -lpthread ../../webthree-helpers/utils/libscrypt/libscrypt.so ../../webthree-helpers/utils/secp256k1/libsecp256k1.a ../../../build-gmp/gmp/mic/lib/libgmp.a ../../../../boost_1_55_0/build-mic/lib/libboost_regex.so ../libethash/libethash.so ../../../../cryptopp-5.6.2/build-mic/lib/libcryptopp.a -Wl,-rpath,/home/json-rpc-cpp-0.5.0/build-mic/lib:/home/mpss-3.7/k1om/usr/lib64:/home/boost_1_55_0/build-mic/lib:/home/libethereum-buildbot/build-mic/libethereum/libethashseal:/home/libethereum-buildbot/build-mic/libethereum/libethereum:/home/libethereum-buildbot/build-mic/libethereum/libevm:/home/libethereum-buildbot/build-mic/libethereum/libethcore:/home/libethereum-buildbot/build-mic/libethereum/libevmcore:/home/libethereum-buildbot/build-mic/libweb3core/libp2p:/home/libethereum-buildbot/build-mic/libweb3core/libdevcrypto:/home/libethereum-buildbot/build-mic/libweb3core/libdevcore:/home/libethereum-buildbot/build-mic/webthree-helpers/utils/libscrypt:/home/libethereum-buildbot/build-mic/libethereum/libethash:/opt/intel/compilers_and_libraries_2016.2.181/linux/compiler/lib/intel64_lin_mic/:/home/mpss-3.7/k1om/lib64/
../../libweb3core/libp2p/libp2p.so: undefined reference to `CryptoPP::AlignedDeallocate(void*)'
../../libweb3core/libp2p/libp2p.so: undefined reference to `CryptoPP::AlignedAllocate(unsigned long)'
[root@coast77 ethminer]# 
```

**Comment by ghost** [comment]
Hello again @bobsummerwill @chriseth,

Resolved the CryptoPP references via additional defines in the CXXFLAGS of libethereum and libweb3core.
`CXXFLAGS="-std=c++11 -mmic -DCRYPTOPP_DISABLE_ASM -DCRYPTOPP_DISABLE_SSE2 -DCRYPTOPP_DISABLE_SSSE3"`

The baseline for Xeon Phi 5110P is ~15-16 MHs @ ~160W.
![k1om_ethminer](https://cloud.githubusercontent.com/assets/7040493/14926224/1bd6566c-0e19-11e6-9993-9ca7bcfb3bde.png)

Can we please merge the follow two pull requests?
https://github.com/ethereum/libethereum/pull/224
https://github.com/ethereum/libweb3core/pull/68

They are required for Xeon Phi operation.
I plan to move on to intrinsic optimizations for k1om architecture.

**Comment by bobsummerwill** [comment]
CryptoPP is a repeated source of pain for us.   Glad you found a workaround.
We plan to remove it at some stage ... https://github.com/ethereum/webthree-umbrella/issues/103.

@chriseth My inclination is to merge these changes as-is.   They are quite small warts, and cleanly ifdefed.   We aren't taking on "ownership" for this support by upstreaming, any more than we have "ownership" for Debian, openSUSE, Tizen, Ubuntu Touch, Sailfish, Raspberry Pi, etc.    We're just accepting patches, and ensuring they don't break any of our existing configs.    The ifdefs aren't lovely, but they're not the worst thing in the world either.

@mancoast has put in a lot of effort working with us here.

**Comment by bobsummerwill** [comment]
@chriseth [said on Gitter](https://gitter.im/ethereum/cpp-ethereum-development?at=5728e4106871c4a646c16856) ... they are horrible, I don't think that they catch all cases and it will break all the time because we do not have an intel compiler as part of our CI (greg already broke them in his PR, for example)
we should dig deeper why intel does not work at that point and fix it at boost::multiprecision and not at the point where boost::multiprecision is used.

So I have just mailed the Boost development mailing list about this.    I will link the post when it shows up in their web archive at http://lists.boost.org/Archives/boost/2016/05/date.php.

**Comment by bobsummerwill** [comment]
My post - http://lists.boost.org/Archives/boost/2016/05/229429.php
Reply just arrived via e-mail and will be in the archive soon, and I will link it here when it's up.

**Comment by jzmaddock** [comment]
I explained this on the Boost mailing list, but in essence:
- The code is relying on a C++11 feature "explicit conversion operators".
- _Implicit _narrowing conversions are always disallowed in Boost.Multiprecision.  No exceptions, no differences by compiler.
- _Explicit_ narrowing conversions are permitted if the compiler supports it, otherwise there is the convert_to member function which does the same thing.
- The presence of BOOST_MP_NO_CXX11_EXPLICIT_CONVERSION_OPERATORS can be used to detect whether Boost.Multiprecision supports explicit conversion operators or not.
- Early implementations of this feature were too buggy to compile Boost.Multiprecision with this feature turned on.  Examples include some Intel versions and all current Oracle C++ versions.  Recent GCC, Clang and MSVC releases are all fine.

Now for the specific case of Intel C++, if all of the following are true then the existing etherium code should work fine:
- Boost-1.60 or later.
- Intel 15 or later.
- C++11 support turned on on the command line.
- Intel is emulating GCC-4.5 or later.

I realise that sounds like a lot of "ifs" but they should really all be true anyway ;)

HTH, John Maddock.

**Comment by ghost** [comment]
Fantastic! I will upgrade boost and confirm compilation.

**Comment by ghost** [comment]
confirmed resolved

**Comment by gcolvin** [comment]
Thanks for clearing that up John @jzmaddock
And thanks for sticking with this so long @mancoast

**Comment by bobsummerwill** [comment]
Thanks for your magnificently detailed reply, @jzmaddock!    Much appreciated.

And thanks indeed for your long and noble work on getting this going, @mancoast!     I am going to reopen this issue, though, because we aren't quite done yet.

After all of this work, I want to ensure that other people are able to build `eth` using the Intel compiler as well, and we're not quite there yet.   I want to lock all of this learning and effort in :-)

Our CMake compiler settings don't have any flows for Intel compiler.

https://github.com/ethereum/webthree-helpers/blob/develop/cmake/EthCompilerSettings.cmake

Did you just have some local edits for this, @mancoast?   Or command-line overrides?   What settings did you use?

I would like us to have instructions for Intel compiler builds to add to http://www.ethdocs.org/en/latest/ethereum-clients/cpp-ethereum/building-from-source/index.html.    What elements of your local setup are we missing, @mancoast?    What instructions do we need to add to get random people in a position where they can replicate your results?

Also, in terms of the specific version requirements, etc. we should get those encoded into the CMake files as well, within that new "if Intel" conditional.

**Comment by bobsummerwill** [comment]
Oh yes, one more thing.

If we don't need the `convert_to<>` changes anymore, I would like to undo them, because they are proving to be a bit of a merge nightmare with a major set of refactorings which @gcolvin is working on in the same codebase.     If they were winning us something we would probably just leave them in, but given that they are causing us pain AND they appear no longer to be required, reverting them back out seems best.

More generally, it looks like our codebase is a pretty horrible mess of C-style casts in general, but that is another story entirely :-)

**Comment by bobsummerwill** [comment]
So I think that would mean reverting the following:
- https://github.com/ethereum/libethereum/pull/224/files
- https://github.com/ethereum/libethereum/pull/223/files (most of, but not 100%)
- https://github.com/ethereum/libweb3core/pull/68/files
- https://github.com/ethereum/libweb3core/pull/67/files (most of, but not 100%)

It is probably best if YOU actually create PRs for these, eh, @mancoast, because you can verify that we don't over-strip in the process, and re-break your Intel builds.

BTW - Are you able to build ALL the executables now?

**Comment by ghost** [comment]
Revert convert_to usage:
https://github.com/ethereum/libethereum/pull/252
https://github.com/ethereum/libweb3core/pull/76

**Comment by ghost** [comment]
Intel x64 cmake command line:

```
# Build x64 #
USRINC="-I$SRC/microhttpd-0.9.28/build/include/ -I$SRC/json-rpc-cpp-0.5.0/build/gen/jsonrpccpp/common/ -I$SRC/json-rpc-cpp-0.5.0/build/gen/"
cd $BUILDTREE \
&& CC=icc CXX=icpc \
CXXFLAGS="-std=c++11 -O2 -g -DETH_RELEASE -pedantic -DSHAREDLIB -fPIC $USRINC" \
CFLAGS="-02 -g" \
BOOST_ROOT=$SRC/boost_1_60_0 \
BOOST_LIBRARYDIR=$SRC/boost_1_60_0/build/lib \
Boost_NO_SYSTEM_PATHS=ON \
cmake3 \
-DGMP_INCLUDE_DIR=$SRC/gmp-6.1.0/build/ \
-DGMP_LIBRARY=$SRC/gmp-6.1.0/build/lib/libgmp.a \
-DJSONCPP_INCLUDE_DIR=$SRC/jsoncpp-1.6.2/include/ \
-DJSONCPP_LIBRARY=$SRC/jsoncpp-1.6.2/build/src/lib_json/libjsoncpp.a \
-DLEVELDB_INCLUDE_DIR=$SRC/leveldb-1.3/include/ \
-DLEVELDB_LIBRARY=$SRC/leveldb-1.3/build/libleveldb.so \
-DCRYPTOPP_INCLUDE_DIR=$SRC/cryptopp-5.6.2/build/include/ \
-DCRYPTOPP_LIBRARY=$SRC/cryptopp-5.6.2/build/lib/libcryptopp.a \
-DJSON_RPC_CPP_INCLUDE_DIR=$SRC/json-rpc-cpp-0.5.0/src/ \
-DJSON_RPC_CPP_COMMON_LIBRARY=$SRC/json-rpc-cpp-0.5.0/build/lib/libjsonrpccpp-common.so \
-DJSON_RPC_CPP_SERVER_LIBRARY=$SRC/json-rpc-cpp-0.5.0/build/lib/libjsonrpccpp-server.so \
-DJSON_RPC_CPP_CLIENT_LIBRARY=$SRC/json-rpc-cpp-0.5.0/build/lib/libjsonrpccpp-client.so \
-DCURL_INCLUDE_DIR=/usr/include/ \
-DCURL_LIBRARY=/usr/lib64/libcurl.so \
-DFATDB=OFF \
-DGUI=OFF \
-DTESTS=OFF \
-DETHASHCL=OFF \
-DEVMJIT=OFF \
-DSOLIDITY=OFF \
-DMINIUPNPC=OFF \
-G "Eclipse CDT4 - Unix Makefiles" \
$SRC/
cd $BUILDTREE && make VERBOSE=1
```

CFLAGS="-02 -g" \ <-- incorrect?

webthree-umbrella\CMakeLists.txt:
`link_directories(${USRLIB_LINK_DIR})`

**Comment by bobsummerwill** [comment]
Thanks, @mancoast  - I have merged the convert_to reverts.

@gcolvin - Please could you do some diffs, and see how your content is now tracking against HEAD of develop for libethereum?

**Comment by chriseth** [comment]
Can this be closed?

**Comment by ghost** [comment]
everything functions on my end.

**Comment by bobsummerwill** [comment]
I've not closed it yet because it still needs documentation.   On my plate!   Will get to it soon.

**Comment by bobsummerwill** [comment]
Hey @mancoast!

Just doing release notes for cpp-ethereum-v1.2.6.   Would you like to be credited with a real name?

Also, whether or not you stay pseudonymous, please can you add an e-mail address which you can be contacted at to http://piratepad.net/g9A0NTQjcI?   Thanks!    That is for https://github.com/ethereum/webthree-umbrella/issues/530

**Comment by ghost** [comment]
all updated, thanks

**Comment by bobsummerwill** [comment]
Hello ROBERT!   Thanks :-)

**Comment by bobsummerwill** [comment]
Moved to https://github.com/ethereum/cpp-ethereum/issues/3219, but not all comments came across cleanly.

**Comment by SKlayer** [comment]
@bobsummerwill  Hello ,i tried to do same thing to compilie to my Xeon Phi. but the intel compiler is not free to student (the XE not include ICC)..... so could you pelease give me a binary code to me? Thank you very much...  My email :1919962626@qq.com 感激不尽
